### PR TITLE
feat: NemotronH PP/EP SFT integration + fixed-length SQuAD supervision fix

### PR DESCRIPTION
**Pull request by jasont314**
# What does this PR do ?

Enable stable NemotronH PP+EP SFT with fixed-length SQuAD by integrating the EP runtime path, hardening PP/FSDP synchronization/schedule behavior, and fixing label-token supervision masking so training no longer degenerates to NaN/zero-label-token steps.

# Changelog

- Add NemotronH EP runtime integration and safety guards in distributed parallelization paths.
- Add EP shard mesh null-guard behavior and mesh-aware EP axis propagation in training setup.
- Add PP schedule robustness updates:
  - explicit invalid-style error handling,
  - guarded `NEMOAUTOMODEL_PP_SKIP_OUTPUT_MERGE` patching with compatibility checks,
  - safer underfill handling interactions in tests.
- Normalize AutoPipeline device handling (`torch.device | int | str` -> `torch.device`) and update call sites.
- Fix FSDP2 divisibility error text to match actual `tp * cp * pp` logic.
- Make warning filtering opt-in via env (`NEMOAUTOMODEL_FILTER_WARNINGS`) for better debugging visibility.
- Fix fixed-length SQuAD supervision path:
  - truncation-aware prompt/answer masking,
  - fixed-length truncation behavior that preserves answer supervision tokens,
  - avoid all-masked label cases (`num_label_tokens == 0`).
- Add optimized SQuAD PP+EP config:
  - `examples/llm_finetune/nemotron/nemotron_nano_v3_pp_ep_squad.yaml`
- Add patch documentation:
  - `NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md`
- Add/update unit tests for:
  - PP schedule/style/skip-merge behavior,
  - AutoPipeline device normalization,
  - EP shard-axis + missing `moe_mesh` safety.
- Include SFT training artifacts for baseline vs optimized throughput comparison:
  - `checkpoints/baseline_training.jsonl`
  - `checkpoints/optimized_training.jsonl`

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to: N/A
- Validation run:
  - `python -m pytest -q tests/unit_tests/distributed/pipelining/test_functional.py tests/unit_tests/distributed/pipelining/test_autopipeline.py tests/unit_tests/moe/test_parallelizer.py`
  - Result: `79 passed, 5 skipped, 6 warnings`

## Acknowledgements
Thanks to collaborators:
- Nazar Ospanov
- Zoir Imomaliev
- Sanjay Adhikesaven

### CODE REVIEW

File: NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+NEMOTRONH_EP_PHYSICAL_PARTITION=1
+NEMOTRONH_EP_SYNC_INACTIVE_EXPERTS=1
+NEMOTRONH_EP_EXPERT_RESHARD_AFTER_FORWARD=0
+NEMOAUTOMODEL_PP_SKIP_OUTPUT_MERGE=1
+```
+
+Meaning of each variable:
+- `TORCH_NCCL_USE_COMM_NONBLOCKING=1`: enables NCCL non-blocking error handling to reduce hard hangs.
+- `PYTORCH_ALLOC_CONF=expandable_segments:True`: reduces allocator fragmentation under large transient GPU allocations.
+- `NEMOTRONH_EP_USE_DEEPEP_DISPATCH=1`: uses DeepEP token dispatch path for EP.
```
Comment (akoumpa) [inline NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md]: Thank you so much for this PR @jasont314 , looks really great.

I was thinking we may want to have these env variables as regular variables in the yaml file instead? Please let me know what you think!

Comment (jasont314) [reply NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md]: Sounds good, just re-pushed the changes!
Implemented: moved runtime env knobs into YAML (dist_env) with mapping applied in train_ft/init_utils, updated config + patch notes, and validated behavior.

Comment (akoumpa) [reply NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md]: Hi @jasont314 thanks a lot for that, one more ask can you rebase to latest main since it now has a lot of conflicts? thank you.

Comment (jasont314) [reply NEMOTRON_PP_EP_SQUAD_PATCH_NOTES.md]: Hi @akoumpa 
I just rebased to latest main, resolved conflicts, and kept the PP/EP + fixed-length SQuAD changes intact.
Also reran the targeted tests (109 passed, 5 skipped) and a quick PP/EP SQuAD successful test run.
Could you please re-review when you get a chance?

**Comment by akoumpa** [comment]
Hi @jasont314 , yes, please take a look at the ci tests failing (unit tests + lint) and the code conflict. Unfortunately, i cannot push to your branch to help with these. Please let me know when ready to trigger ci. thank you.

**Comment by akoumpa** [comment]
Hi @jasont314, I owe you an apology, this sat too long on my end, and now it has significant conflicts on top.

I realize it's a lot to ask, but would you be willing to rebase and split it into smaller PRs? It touches the whole codebase and landing it in chunks would be much safer to review and merge. Totally understand if the timing no longer works for you. 

Please let me know what you'd like to do. Thank you.

**Comment by jasont314** [comment]
Hi @akoumpa , I might not be able to re-implement everything immediately as I am currently finalizing a research project. If I have time after I finish, I will be able to return to this issue. However, I only have access to GPUs for another month or so. I apologize for the timing.

**Comment by akoumpa** [comment]
Hi @jasont314 , thanks for letting me know. I'll keep this open and please feel free to pick this up as time/bandwidth permit. Thank you again for the contribution!
